public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/LabelsEditor.tsx

```diff
@@ -32,7 +33,8 @@ export function LabelsEditor(props: Props) {
       <CodeEditor
         value={props.value}
         language={langId}
-        onBlur={props.onChange}
+        onChange={props.onChange}
```

Comment:
This was necessary to get proper updates as the user is typing, but made the editor slower initially because changes to the query state object caused a lot of things to re-render. 

This is the reason for the bigger structural changes done to `QueryEditor` which make things run smooth again. I am curious if there are other more elegant solutions for this.

Reply:
If `onChange` is called "too often", we could consider [debouncing](https://lodash.com/docs/4.17.15#debounce).